File: c/problem_0015.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "./lib/mathlib.h"

/**
 * @see https://projecteuler.net:
 *
 * Lattice paths
 * Problem 15
 *
 * Starting in the top left corner of a 2×2 grid, and only being able to move to the right and down, there are exactly 6 routes to the bottom right corner.
 *
 * How many such routes are there through a 20×20 grid?
 */

unsigned long calcRoutes(int size, int startX, int startY)
{
    unsigned long sum = 0;
    if (startX == size-1 && startY == size-1) {
        return 1;
    }
    // Can I go right?
    if (startX < size - 1) {
        sum += calcRoutes(size, startX+1, startY);
    }
    // Can I go down?
    if (startY < size - 1) {
        sum += calcRoutes(size, startX, startY+1);
    }
    return sum;
}

#define MAP_SIZE 21
int main(void)
{
    unsigned long nrOfRoutes = calcRoutes(MAP_SIZE, 0, 0);
    printf("Nr of routes in a %ix%i grid: %lu\n", MAP_SIZE - 1, MAP_SIZE - 1, nrOfRoutes);
}

File: haskell/problem_0015.hs
{--------------------------------------------------------------------
<p>Starting in the top left corner of a $2 \times 2$ grid, and only being able to move to the right and down, there are exactly $6$ routes to the bottom right corner.</p>
<div class="center">
<img src="https://projecteuler.net/resources/images/0015.png?1678992052" class="dark_img" alt=""></div>
<p>How many such routes are there through a $20 \times 20$ grid?</p>
---------------------------------------------------------------------}

import Data.Map.Strict qualified as Map
import EulerLib qualified as E

type CoordCountMap = Map.Map (Int, Int) Int

-- | count the routes for a given coordinate, working with and returning a memoization
-- | map of already visited routes.
countRoutes :: (Int, Int) -> (Int, Int) -> CoordCountMap -> (Int, CoordCountMap)
countRoutes (x, y) (maxX, maxY) cMap
  -- final position: return 1
  | x == maxX && y == maxY = (1, cMap)
  -- out-of-bounds position: return 0 (no route)
  | x > maxX || y > maxY = (0, cMap)
  | otherwise = case actCount of
      -- already here once: just return the route count
      Just cnt -> (cnt, cMap)
      -- not yet here: sum the right + down route count and store:
      Nothing -> (cntAll, finalMap)
        where
          (cnt1, map1) = countRoutes (x, y + 1) (maxX, maxY) cMap
          (cnt2, map2) = countRoutes (x + 1, y) (maxX, maxY) map1
          cntAll = cnt1 + cnt2
          finalMap = Map.insert (x, y) cntAll map2
  where
    actCount = Map.lookup (x, y) cMap

main = do
  let eulerProblem = 15
      (solution, _) = countRoutes (0, 0) (20, 20) Map.empty
  E.printProblemSolution eulerProblem solution
